File: backend/app/models/recipe.rb
class Recipe < ApplicationRecord
  has_many :recipe_ingredients, dependent: :destroy
  has_many :ingredients, through: :recipe_ingredients

  enum :meal_type,  { breakfast: 0, lunch: 1, dinner: 2, snack: 3 }
  enum :difficulty, { easy: 1, normal: 2, hard: 3 }

  validates :name,       presence: true, length: { maximum: 100 }
  validates :meal_type,  presence: true
  validates :difficulty, presence: true

  def nutrition_totals
    NutritionCalculator.for_recipe(self).call
  end

  def ingredient_ids_set
    ingredient_ids.to_set
  end

  def as_api_hash
    nutrition = NutritionCalculator.for_recipe(self).call
    {
      id:                   id,
      name:                 name,
      description:          description,
      meal_type:            meal_type,
      cooking_time_minutes: cooking_time_minutes,
      difficulty:           difficulty,
      instructions:         instructions&.split("\n") || [],
      tags:                 tags&.split(",") || [],
      ingredients:          recipe_ingredients.map do |ri|
        { id: ri.ingredient_id, name: ri.ingredient.name, amount_g: ri.amount_g.to_f }
      end,
      nutrition:            nutrition.to_h
    }
  end
end

File: frontend/src/types/recipe.ts
export type Difficulty = 'easy' | 'normal' | 'hard'
export type MealType = 'breakfast' | 'lunch' | 'dinner' | 'snack'

export type RecipeIngredient = {
  id: number
  name: string
  amount_g: number
}

// NutritionCalculator::Result が Float で返すため number
export type Nutrition = {
  protein_g: number
  fat_g: number
  carbohydrate_g: number
  sodium_g: number
  calories_kcal: number
}

export type Recipe = {
  id: number
  name: string
  description: string | null
  meal_type: MealType
  cooking_time_minutes: number | null
  difficulty: Difficulty
  instructions: string[]
  tags: string[]
  ingredients: RecipeIngredient[]
  nutrition: Nutrition
}

export type RecipeSearchResponse = {
  matched_recipes: Recipe[]
  count: number
}

export type RecipeSearchInput = {
  ingredient_ids: number[]
}
